### PR TITLE
Add configurable product stock counter with condition

### LINKED ISSUE #2122: Out of stock products are displayed on the Category page
**Description**:
Some out of stock products are displayed on the Category page

**Expected behavior**:
Out of stock products **_should not_** be displayed on the Category page

**Screenshots, Video, Logs**
Screenshot from 2021-01-26 09-30-31

**To Reproduce**:
1. Go to [the women category](https://scandipwapmrev.indvp.com/default/n-comma-casual-identity-quilted-jacket.html)
2. Leaf through pages until you see `Comma Casual Identity Quilted Jacket`
3. Open both in new tabs
4. One of them will result 404

### PR DESCRIPTION
**Pull request by winniepukki**
Fixes scandipwa/scandipwa#2122

### CODE REVIEW

**Review by alfredsgenkins — APPROVED** [review]
Seems OK, but this logic becomes more and more complex.

You handled the simple and configurable cases - what about bundle items? Also, does this respects the "disabled" stock tracking (or how it's correctly named)? @lianastaskevica did QA check all above cases?

**Comment by lianastaskevica** [comment]
> Seems OK, but this logic becomes more and more complex.
> 
> You handled the simple and configurable cases - what about bundle items? Also, does this respects the "disabled" stock tracking (or how it's correctly named)? @lianastaskevica did QA check all above cases?

@alfredsgenkins we cannot QA BE changes. @winniepukki please test your solution for all product types and update us here on the status.

**Comment by winniepukki** [comment]
@alfredsgenkins @lianastaskevica failed for the Bundle type. Everything else that has been tested with my local set-up passed with all possible scenarios

File: src/Model/Resolver/EntityUrl.php [hunk 1]
```diff
@@ -239,4 +295,54 @@ private function sanitizeType(string $type) : string
...
+    }
+
+    private function getProductType($id): ?string
+    {
+        try {
+            return $this->productRepository->getById($id)->getTypeId();
+        } catch (NoSuchEntityException $e) {
+            return null;
+        }
+    }
```
Comment (alfredsgenkins) [inline src/Model/Resolver/EntityUrl.php]: This triggers the load of product from DB every time, do we really need it?

File: src/Model/Resolver/EntityUrl.php [hunk 2]
```diff
@@ -141,10 +185,22 @@ public function resolve(
...
+                    $isInStock = $this->getConfigurableProductStockState($product);
+                } elseif ($productType === 'bundle') {
+                    $isInStock = $this->getBundleProductsStockState($this->getBundleProductChildrenItems($product, $id));
+                } else {
+                    try {
+                        $isInStock = $this->stockItemRepository->get($id)->getIsInStock();
+                    } catch (NoSuchEntityException $e) {
+                        // Ignoring error is safe
+                    }
                 }
```
Comment (alfredsgenkins) [inline src/Model/Resolver/EntityUrl.php]: Now this is dirty, can you please try writing that in some less **iffy** way? switch? Maybe also move the product type obtaining into separate function?

**Review by alfredsgenkins — CHANGES_REQUESTED** [review]

**Pull request merged**